keps/sig-network/3178-iptables-cleanup/README.md

```diff
@@ -0,0 +1,687 @@
+processing than kube-proxy does, they could fairly easily be rewritten
+to not use the packet mark, and just have slightly-redundant
+`PREROUTING` and `POSTROUTING` rules instead.
+
+[CNI portmap plugin]: https://github.com/containernetworking/plugins/tree/master/plugins/meta/portmap
+[CRI-O]: https://github.com/cri-o/cri-o
+
+#### `iptables-wrapper`
+
+Another problem is the [iptables-wrapper script] for detecting whether
+the system is using iptables-legacy or iptables-nft. Currently it
+assumes that kubelet will always have created _some_ iptables chains
+before the wrapper script runs, and so it can decide which iptables
+backend to use based on which one kubelet used. If kubelet stops
+creating chains, then iptables-wrapper may find itself in a situation
+where there are no chains in either set of tables.
+
+```
+<<[UNRESOLVED iptables-wrapper-help ]>>
+
+It might make sense to have kubelet continue to create a single
+IPTables chain that exists for no purpose other than to indicate to
+pod-based external components whether the system is using
+iptables-legacy or iptables-nft.
+
+<<[/UNRESOLVED]>>
+```
+
+[iptables-wrapper script]: https://github.com/kubernetes-sigs/iptables-wrappers
+
+## Design Details
+
+### Implementation
+
+We cannot remove `KUBE-MARK-DROP` from kubelet until we know that
+kubelet cannot possibly be running against a version of kube-proxy
+that requires it.
+
+Thus, the process will be:
+
+#### Alpha
+
+We will document, in the appropriate places, the fact that external
+components should not be assuming the existence of Kubernetes's
+IPTables chains. (This probably includes a release note and a blog
+post.)
+
+Kube-proxy will be updated to not use `KUBE-MARK-DROP`, as described
+above. (This change is unconditional; it is _not_ feature-gated.) We
+should also ensure that kpng gets updated.
+
+Kubelet's behavior will not change by default, but if you enable the
+`IPTablesCleanup` feature gate, then:
+
+  1. It will stop creating `KUBE-MARK-DROP`, `KUBE-MARK-MASQ`,
+     `KUBE-FIREWALL`, and `KUBE-POSTROUTING`.
+
+  2. It will delete the `KUBE-MARK-DROP` and `KUBE-FIREWALL` chains,
+     if they exist.
+
+  3. It will warn that the `--iptables-masquerade-bit` and
+     `--iptables-drop-bit` flags are deprecated and have no effect.
+
+(Importantly, those kubelet flags will _not_ claim to be deprecated
+when the feature gate is disabled, because in that case it is
+important that if the user was previously overriding
+`--iptables-masquerade-bit`, that they keep overriding it in both
+kubelet and kube-proxy for as long as they are both redundantly
+creating the chains.)
+
+#### Beta
+
+The behavior is the same as alpha, except that the feature gate is
+enabled by default
+
+As long as we wait 2 releases between Alpha and Beta, then it is
+guaranteed that when the user upgrades to Beta, the currently-running
+kube-proxy will be one that does not require the `KUBE-MARK-DROP`
```

Comment:
What about people running other proxies?  E.g. cilium - it seems unlikely that literally cilium depends on iptables rules, but I've seen worse.  And there ARE other implementations out there..

Reply:
Part of the work for the Alpha release is:

> We will document, in the appropriate places, the fact that external
components should not be assuming the existence of Kubernetes's
IPTables chains. (This probably includes a release note and a blog
post.)

(Though really this should be done well before the actual alpha release date, to give third-party developers time to get an updated release out in parallel with the `IPTablesCleanup` alpha.)

Also, the graduation criteria for Beta includes:

> - If the removal of the IPTables chain creation from kubelet has been
  found to cause problems with third-party components, then the SIG
  must agree that these problems can be addressed with release notes
  and do not justify delaying Beta. (eg, a bug involving a component
  which is not widely used, or which can be worked around by upgrading
  to a newer version of that component, or changing its
  configuration.)

(That's kind of badly-worded because this section is "criteria for proceeding", not "criteria for delaying", but the intent here is "if the SIG thinks the world isn't ready for this KEP to go to Beta, then it won't go to Beta".)